In [1]:
import os
import sys
import glob
import time
import pandas as pd
import json
import datetime

# Data Paths

In [2]:
data_path = "/DATA/7A/xblock/"
df_path = "/DATA/7A/assets/dataframes/"

#
erc20_transactions_path = "ERC20Transactions/uncompressed"
erc721_transactions_path = "ERC721Transactions/uncompressed"
internal_transactions_path = "InternalTransactions/uncompressed"
normal_transactions_path = "NormalTransactions/uncompressed"

#token information
token_info_path = "TokenInfo/uncompressed/"
erc20_filename = "ERC20TokenInfo.csv"
erc721_filename = "ERC721TokenInfo.csv"

#dataframes
internalTransactions_filename = "internalTransactions.df"
erc20Transactions_filename = "erc20Transactions.df"
erc721Transactions_filename = "erc721Transactions.df"


# Get Token Information and Transactions

##  get erc721 token information

In [3]:
#get erc721 token information
erc721InfoDF = pd.read_csv(data_path + token_info_path + erc721_filename, error_bad_lines=False)
erc721InfoDF = erc721InfoDF[erc721InfoDF['decimal'] != "None"].astype({"decimal": float})
print(erc721InfoDF[erc721InfoDF["address"] == "0xb932a70a57673d89f4acffbe830e8ed7f75fb9e0"])
print(erc721InfoDF.shape[0])
erc721InfoDF.head()

                                      address       name symbol  decimal
8  0xb932a70a57673d89f4acffbe830e8ed7f75fb9e0  SuperRare   SUPR  16651.0
7691


address              name  symbol  \
0  0x27b4bc90fbe56f02ef50f2e2f79d7813aa8941a7         gego.dego    GEGO   
1  0x3f9d7b15763f57d7f5015901eaccca27ce996bb2        Polysphere  PLYSPR   
2  0x06012c8cf97bead5deae237070f9587f8e7a266d     CryptoKitties      CK   
3  0xbbe7aeb729e65e3854a30d583cee6a28434eb841        Polysphere    POLY   
4  0x2fb5d7dda4f1f20f974a0fdd547c38674e8d940c  KnightStory Item    MyKI   

     decimal  
0    54084.0  
1        5.0  
2  1996597.0  
3        0.0  
4    13250.0

In [25]:
print(erc721InfoDF[erc721InfoDF["address"] == "0x20f7a3ddf244dc9299975b4da1c39f8d5d75f05a"])
erc721InfoDF[erc721InfoDF["name"] == "IdeaMarket"][["address"]].values.ravel()

Empty DataFrame
Columns: [address, name, symbol, decimal]
Index: []


array([], dtype=object)

##  get erc20 token information

In [30]:
#get erc20 token information
#line 154118 in this file is problematic for some reason
erc20InfoDF = pd.read_csv(data_path + token_info_path + erc20_filename, error_bad_lines=False)
print(erc20InfoDF[erc20InfoDF["address"] == "0x3Ee07a2197E6EE3590D8f6408915A7707AAB8094"])
print(erc20InfoDF[erc20InfoDF["name"] == "Den"])
print(erc20InfoDF.shape[0])
erc20InfoDF.head()

Empty DataFrame
Columns: [address, name, symbol, totalSupply, decimal]
Index: []
                                           address name symbol  \
144508  0xd384f231f5808509586ba570a0709a38a0f50991  Den    Den   

                     totalSupply decimal  
144508  100000000000000000000000      18  
226250


b'Skipping line 154118: expected 5 fields, saw 7\n'


address                  name symbol  \
0  0x6b0359f95796327475ad4f12ae4e1047c3a67fa3         FinFine Token    FFT   
1  0x12d79c345cac7b050a5ff0797b5a607e254c73f5                Buytex    BUX   
2  0x6bc1f3a1ae56231dbb64d3e82e070857eae86045                xensor    XSR   
3  0x10086399dd8c1e3de736724af52587a2044c9fa2  The Midas Touch Gold   TMTG   
4  0xdac17f958d2ee523a2206206994597c13d831ec7            Tether USD   USDT   

                    totalSupply decimal  
0                  214285500000       3  
1                 9999999999999       4  
2  4800000000000000000000000000      18  
3  9161139666073037500000000000      18  
4             20817977739617189       6

## read all internal transactions

In [4]:
#all_files = glob.glob(data_path + internal_transactions_path + "/*.csv")
#
#li = []
#
#for filename in all_files:
#    start = time.time()
#    print("Reading file", filename, "...", end='', flush=True)
#    df = pd.read_csv(filename, index_col=None, header=0)
#    li.append(df)
#    end = time.time()
#    print(" took", end - start, "seconds.")
#    #break #read just the first file
#internalTransactionsDF = pd.concat(li, axis=0, ignore_index=True)

s = time.time()
internalTransactionsDF = pd.read_feather(df_path + internalTransactions_filename)
e = time.time()
print("Took", e - s, " seconds.")
print(internalTransactionsDF.shape[0])
internalTransactionsDF.head(10)

NameError: name 'df_path' is not defined

## read all erc20 transactions

In [19]:
#all_files = glob.glob(data_path + erc20_transactions_path + "/*.csv")
#
#li = []
#
#for filename in all_files:
#    start = time.time()
#    print("Reading file", filename, "...", end='', flush=True)
#    df = pd.read_csv(filename, index_col=None, header=0)
#    li.append(df)
#    end = time.time()
#    print(" took", end - start, "seconds.")
#    #break #read just the first file
#erc20TransactionsDF = pd.concat(li, axis=0, ignore_index=True)

s = time.time()
erc20TransactionsDF = pd.read_feather(df_path + erc20Transactions_filename)
e = time.time()
print("Took", e - s, " seconds.")
print(erc20TransactionsDF.shape[0])
erc20TransactionsDF.head()

Took 703.6594405174255  seconds.
601955094


blockNumber   timestamp                                    transactionHash  \
0       462680  1446203330  0x806172da2073b5c3d961a8c613b7e469766ad19c6220...   
1       462683  1446203369  0xd5c17b09dcf6e5daa0566d7d2a8245f10a755ecfbf34...   
2       462696  1446203551  0x530e047cccef3ea1f78bcdb629661fabe193ba87cd82...   
3       477958  1446458222  0x0d58cd643c38a8951ee26a49399d4d5682fab74e96a9...   
4       483333  1446552343  0xd628780ba231cefe6a4f6c3da3b683b16f6151dc9753...   

                                 tokenAddress  \
0  0xe6ee69495b571e1042f760d7f34009164aff87a2   
1  0xe6ee69495b571e1042f760d7f34009164aff87a2   
2  0xe6ee69495b571e1042f760d7f34009164aff87a2   
3  0xe6ee69495b571e1042f760d7f34009164aff87a2   
4  0xf4eced2f682ce333f96f2d8966c613ded8fc95dd   

                                         from  \
0  0xb1a2b43a7433dd150bb82227ed519cd6b142d382   
1  0xb1a2b43a7433dd150bb82227ed519cd6b142d382   
2  0xb1a2b43a7433dd150bb82227ed519cd6b142d382   
3  0xb1a2b43a7433dd150bb82227ed519cd6b142d382   
4  0x9b22a80d5c7b3374a05b446081f97d0a34079e7f   

                                           to  fromIsContract  toIsContract  
0  0x9b22a80d5c7b3374a05b446081f97d0a34079e7f               0             0  
1  0x9b22a80d5c7b3374a05b446081f97d0a34079e7f               0             0  
2  0x9b22a80d5c7b3374a05b446081f97d0a34079e7f               0             0  
3  0xcf8470d3388a198418e02aaf8e37392cfbc18867               0             1  
4  0xd1220a0cf47c7b9be7a2e6ba89f429762e7b9adb               0             0

## read all erc721 transactions

In [4]:
#all_files = glob.glob(data_path + erc721_transactions_path + "/*.csv")
#
#li = []
#
#for filename in all_files:
#    start = time.time()
#    print("Reading file", filename, "...", end="", flush=True)
#    df = pd.read_csv(filename, index_col=None, header=0)
#    li.append(df)
#    end = time.time()
#    print(" took", end - start, "seconds.")
#    #break #read just the first file
#erc721TransactionsDF = pd.concat(li, axis=0, ignore_index=True)

s = time.time()
erc721TransactionsDF = pd.read_feather(df_path + erc721Transactions_filename)
e = time.time()
print("Took", e - s, " seconds.")
print(erc721TransactionsDF.shape[0])
erc721TransactionsDF.head()

Took 12.031283378601074  seconds.
29634892


blockNumber   timestamp                                    transactionHash  \
0      3006260  1484564580  0x9d5fe398fe902975f8508e197ec6c0d0082f8a142a7f...   
1      3053922  1485244952  0xd97c4ddc0f5c0af8c18a093b184de0a8a14e35c87c3f...   
2      3053922  1485244952  0xd97c4ddc0f5c0af8c18a093b184de0a8a14e35c87c3f...   
3      3053922  1485244952  0xd97c4ddc0f5c0af8c18a093b184de0a8a14e35c87c3f...   
4      3132454  1486369652  0x2a7ed459141cc841ab5b376ea64dbc7041b333384d7a...   

                                 tokenAddress  \
0  0xe9436e4cb00f38215a941c7141ee98348fdcce6d   
1  0x55b9a11c2e8351b4ffc7b11561148bfac9977855   
2  0x55b9a11c2e8351b4ffc7b11561148bfac9977855   
3  0x55b9a11c2e8351b4ffc7b11561148bfac9977855   
4  0x55b9a11c2e8351b4ffc7b11561148bfac9977855   

                                         from  \
0  0x002add5778a3d327278a357a4983a57c1894af12   
1  0x121b1badeda482fa25f01c63ad17bbc2701f9165   
2  0x121b1badeda482fa25f01c63ad17bbc2701f9165   
3  0x121b1badeda482fa25f01c63ad17bbc2701f9165   
4  0x01d80e458ae6cd5d2bce90b06ff2d3fa993b65f5   

                                           to  fromIsContract  toIsContract  
0  0xa84ab927943d87e2f14afac62af7283e26cf010e               0             0  
1  0x5d2c24efac49825f682b5d334428f1a5c96acc09               0             1  
2  0x3effc3bb636a66d39ae016399d787d99e1c286a4               0             0  
3  0x5d2c24efac49825f682b5d334428f1a5c96acc09               0             1  
4  0x5d2c24efac49825f682b5d334428f1a5c96acc09               0             1

***

***

***

# Analysis

In [8]:
import numpy as np

## Token Statistics

In [31]:
targetInfoDF = erc721InfoDF
targetDF = erc721TransactionsDF


s = time.time()
#get contract addresses
contractAdresses = targetDF[targetDF["tokenAddress"].isin(targetInfoDF["address"].values)]

#count transactions by tokenAddress 
contractAdresses = contractAdresses.tokenAddress.value_counts()
targetInfoDF['number_of transactions'] = targetInfoDF['address'].map(contractAdresses)
targetInfoDF = targetInfoDF[["name", "number_of transactions"]].groupby(by=["name"]).sum()
targetInfoDF = targetInfoDF.sort_values(by=['number_of transactions'], ascending=False, na_position='last')
e = time.time()
print("Took", e - s, " seconds.")

#token_address = "0xD1CEeeeee83F8bCF3BEDad437202b6154E9F5405"
#token_symbol = "CHZ"
#print(targetInfoDf[targetInfoDf["symbol"] == token_symbol])
#print(targetInfoDf[targetInfoDf["address"] == token_address])
targetInfoDF.head(30)

Took 5.605526447296143  seconds.


number_of transactions
name                                              
Gods Unchained Cards                      15310120
CryptoKitties                              5450248
Gods Unchained                              603051
DozerDoll                                   579632
Codex Record                                523779
LucidSight-MLB-NFT                          477187
Sorare                                      414618
MARBLE-NFT                                  187361
Crypton                                     168897
Decentraland LAND                           163516
HyperDragons                                154783
Crypto stamp Edition 1                      152824
MyCryptoHeroes:Extension                    151279
MyCryptoHeroes:Hero                         144896
gego.dego                                   136046
Servant                                     134214
Spheroid SPACE                              109155
KnightStory Item                            100547
Etherbots                                    91054
Rarible                                      88837
CryptoFlowers                                69623
CryptoAssaultUnit                            59664
Baroque Game                                 59551
MCP Resource                                 56566
EtheremonMonster                             56495
None                                         54089
EtheremonAdventure                           52517
Neon District In-Game Item                   51861
Clovers                                      50960
CryptoBots                                   49756

In [22]:
targetInfoDF.head(50)

number_of transactions
name                                         
Tether USD                         93686246.0
Wrapped Ether                      36040995.0
USD Coin                           13087455.0
None                               10718183.0
More Gold Coin                      8947534.0
Dai Stablecoin                      8080096.0
ChainLink Token                     7652679.0
Livepeer Token                      5638106.0
coToken                             5367787.0
Uniswap V2                          4218841.0
Bancor Network Token                3321288.0
Paxos Standard                      3103713.0
OMGToken                            3093444.0
Alchemy Coin                        3064195.0
Basic Attention Token               2976897.0
Tronix                              2935908.0
Liberty Cash                        2790557.0
Minereum                            2428970.0
Ether Token                         2268795.0
CpcToken                            2220879.0
MSD                                 2113115.0
Maximine Coin                       2046184.0
Uniswap                             1980210.0
BeautyChain                         1925804.0
0x Protocol Token                   1835469.0
StorjToken                          1825611.0
Chi Gastoken by 1inch               1742919.0
BitAnex Coin                        1674519.0
SushiToken                          1665782.0
Pundi X Token                       1632346.0
Esch$Token                          1616611.0
Tepleton                            1593434.0
SiaCashCoin                         1549398.0
HEX                                 1481203.0
WorldUnitedCoins                    1427426.0
0xUniverse                          1422855.0
Wrapped BTC                         1404144.0
Baer Chain                          1383146.0
Kyber Network Crystal               1381999.0
SUM TOKEN                           1367719.0
KickToken                           1332530.0
An Etheal Promo                     1325055.0
NEST                                1299729.0
Enjin Coin                          1234909.0
SuperSkyNet                         1191067.0
TrueUSD                             1188367.0
SBToken                             1173576.0
WENI                                1165375.0
flamingostar                        1161972.0
Compound                            1134644.0

##### Some erc721 tokens have a different decimal and number of transactions. Why is that?

**Different**

CryptoKitties => decimal 1996597, number_of transactions 5434740

MARBLE-NFT => decimal 60452, number of transactions 187361

**Same**

Gods Unchained Cards => decimal 6680552, number_of transactions 6680552

Gods Unchained => decimal 603050, res 603050

Gods Unchained Cards => decimal 580736, res 580736



**Dapps by Type**

Games:
https://sorare.com/
https://cryptodozer.io/
https://godsunchained.com/

Marketplace:
https://rarible.com/

### I believe we need relatively recent data ! (Not sure tho)



## Significance of tokens

In [16]:
erc721InfoDF.head(10)
erc20InfoDF.head(10)

erc721Addresses = erc721InfoDF["address"].values
erc20Addresses = erc20InfoDF["address"].values
tokenAddresses = np.concatenate([erc20Addresses, erc721Addresses])
print(len(tokenAddresses))
tokenAddresses

# token transcation ratio
tokenTransactionRatio = 

233941


array(['0x6b0359f95796327475ad4f12ae4e1047c3a67fa3',
       '0x12d79c345cac7b050a5ff0797b5a607e254c73f5',
       '0x6bc1f3a1ae56231dbb64d3e82e070857eae86045', ...,
       '0xf5c8380f5656f7d82b36dfe5fa2f484839e2fc39',
       '0x6721f071cf1a7d10f5309a896ffa1a724a9a705c',
       '0xf74db21536d2ee9b8428bfee0394586771e13bf0'], dtype=object)